### PR TITLE
Added filter plugin lists_mergeby. Fix #249

### LINKED ISSUE #249: New merge lists plugin from a given key
##### SUMMARY

Given two lists like:

```yaml
list1:
  - name: myname01
    param01: myparam01
  - name: myname02
    param01: myparam02

list2:
  - name: myname01
    param01: myparam03
  - name: myname02
    param02: myparam04
  - name: myname03
    param03: myparam03
```

When doing:

```yaml
"{{ query('mergelist', 'name', list1, list2) }}"
```

The results will be:

```yaml
  - name: myname01
    param01: myparam03
  - name: myname02
    param01: myparam02
    param02: myparam04
  - name: myname03
    param03: myparam03
```

##### ISSUE TYPE
- Feature Idea

##### COMPONENT NAME

`mergelist`
##### ADDITIONAL INFORMATION

I have already developed the solution. I really think it could be really useful in the Ansible Collections General. Please confirm.

```yaml
  - name: Basic usage
    debug:
      msg: "{{ query('mergelist', 'name', list1, list2) }}"
```

### PR DESCRIPTION
**Pull request by vbotka**
##### SUMMARY

See #249

##### ISSUE TYPE

- Feature Pull Request

##### COMPONENT NAME

lists_mergeby
##### ADDITIONAL INFORMATION

```paste below

```

### CODE REVIEW

File: tests/integration/targets/filter_list/tasks/main.yml [hunk 1]
```diff
@@ -0,0 +1,6 @@
+---
+- name: test lists meged by attribute name
```
[the remark this answers is not in the document]
Comment (vbotka) [reply tests/integration/targets/filter_list/tasks/main.yml:2]: Is there a rule that a name shall start with a lower-case letter? I' not aware of such a code of conduct. The developer's documentation [example](https://docs.ansible.com/ansible/latest/dev_guide/developing_modules_documenting.html#examples-block) shows upper-case name! Why do you propose a lower-case name?

File: plugins/filter/list.py [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+from collections import defaultdict
+from operator import itemgetter
+
+
+def lists_mergeby(l1, l2, index):
+    '''Merge lists by attribute index. Example:
+    - debug: msg="{{ l1|lists_mergeby(l2, 'index')|list }}"
+    '''
+    d = defaultdict(dict)
+    for l in (l1, l2):
```
[the remark this answers is not in the document]
Comment (vbotka) [reply plugins/filter/list.py:18]: Right. I'll update PR.

File: tests/integration/targets/filter_list/tasks/main.yml (hunk 1 above)
[the remark this answers is not in the document]
Comment (felixfontein) [reply tests/integration/targets/filter_list/tasks/main.yml:2]: If you look closely, you can see that you used a lower-case name, while I propose to change it to upper-case (as in the dev docs example).

[the remark this answers is not in the document]
Comment (vbotka) [reply tests/integration/targets/filter_list/tasks/main.yml:2]: Right. Let's keep it that way.

**Comment by Andersson007** [comment]
any updates on this?

**Comment by Andersson007** [comment]
to retrigger tests

**Comment by gundalow** [comment]
@imjoseangel @Krakoslabs @lupocito @vbotka What's your view, use filter, or lookup

https://github.com/ansible-collections/community.general/pull/263
https://github.com/ansible-collections/community.general/pull/604

**Comment by dmsimard** [comment]
> @imjoseangel @Krakoslabs @lupocito @vbotka What's your view, use filter, or lookup
> 
> #263
> #604

From a functional point of view, a lookup generally queries something (a system, an application, an API, etc.) and then returns data while a filter takes existing input data and munges it into something else.

With that in mind a filter would be a better fit for this.

File: plugins/filter/list.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (vbotka) [reply plugins/filter/list.py:30]: I've added the testing of types and decided not to test the presence of the index in the lists. Instead, an empty list will be returned if the index is not present. I'd consider this a valid scenario. What do you think?

[the remark this answers is not in the document]
Comment (vbotka) [reply plugins/filter/list.py:31]: I've added the changelog fragment too. It's valid YAML formatted for better readability. Would this fit the purpose?

File: changelogs/fragments/604-lists_mergeby-new-filter.yml [hunk 3]
```diff
@@ -0,0 +1,13 @@
...
+    For example:
+    [{'n': 'n1', 'p1': 'A', 'p2': 'F'},
+     {'n': 'n2', 'p2': 'B'}]|lists_mergeby(
+    [{'n': 'n1', 'p1': 'C'},
+     {'n': 'n2', 'p2': 'D'},
+     {'n': 'n3', 'p3': 'E'}], 'n')|list
+    evaluates to
+    [{'n': 'n1', 'p1': 'C', 'p2': 'F'},
+     {'n': 'n2', 'p2': 'D'},
+     {'n': 'n3', 'p3': 'E'}]"
```
Comment (felixfontein) [inline changelogs/fragments/604-lists_mergeby-new-filter.yml:13]:
```suggestion
  - |
    A new filter ``lists_mergeby`` to merge two lists by an attribute.
    For example:

    .. code-block:: yaml

        [{'n': 'n1', 'p1': 'A', 'p2': 'F'},
         {'n': 'n2', 'p2': 'B'}] | community.general.lists_mergeby(
        [{'n': 'n1', 'p1': 'C'},
         {'n': 'n2', 'p2': 'D'},
         {'n': 'n3', 'p3': 'E'}], 'n') | list

    evaluates to

    .. code-block:: yaml

        [{'n': 'n1', 'p1': 'C', 'p2': 'F'},
         {'n': 'n2', 'p2': 'D'},
         {'n': 'n3', 'p3': 'E'}]

    (https://github.com/ansible-collections/community.general/pull/604).
```

File: plugins/filter/list.py [hunk 4]
```diff
@@ -0,0 +1,47 @@
+# -*- coding: utf-8 -*-
+# Copyright (c) 2020, Vladimir Botka <vbotka@gmail.com>
+# GNU General Public License v3.0+ (see COPYING or https://www.gnu.org/licenses/gpl-3.0.txt)
+
+from __future__ import (absolute_import, division, print_function)
+__metaclass__ = type
+
+from ansible.errors import AnsibleError, AnsibleFilterError
+from ansible.module_utils.six import string_types
```
Comment (felixfontein) [inline plugins/filter/list.py:9]:
```suggestion
from ansible.module_utils.six import string_types
from ansible.module_utils.common._collections_compat import Mapping, Sequence
```

File: plugins/filter/list.py [hunk 5]
```diff
@@ -0,0 +1,47 @@
...
+from collections import defaultdict
+from operator import itemgetter
+
+
+def lists_mergeby(l1, l2, index):
+    ''' merge lists by attribute index. Example:
+        - debug: msg="{{ l1|community.general.lists_mergeby(l2, 'index')|list }}" '''
+
+    if not isinstance(l1, list):
```
Comment (felixfontein) [inline plugins/filter/list.py:18]:
```suggestion
    if not isinstance(l1, Sequence):
```

File: plugins/filter/list.py [hunk 6]
```diff
@@ -0,0 +1,47 @@
...
+        raise AnsibleFilterError('First argument for lists_mergeby must be list. %s is %s' %
+                                 (l1, type(l1)))
+
+    if not isinstance(l2, list):
```
Comment (felixfontein) [inline plugins/filter/list.py:22]:
```suggestion
    if not isinstance(l2, Sequence):
```

File: plugins/filter/list.py [hunk 7]
```diff
@@ -0,0 +1,47 @@
...
+from collections import defaultdict
+from operator import itemgetter
+
+
+def lists_mergeby(l1, l2, index):
+    ''' merge lists by attribute index. Example:
+        - debug: msg="{{ l1|community.general.lists_mergeby(l2, 'index')|list }}" '''
+
+    if not isinstance(l1, list):
+        raise AnsibleFilterError('First argument for lists_mergeby must be list. %s is %s' %
```
Comment (felixfontein) [inline plugins/filter/list.py:19]:
```suggestion
        raise AnsibleFilterError('First argument for community.general.lists_mergeby must be list. %s is %s' %
```

File: plugins/filter/list.py [hunk 8]
```diff
@@ -0,0 +1,47 @@
...
+                                 (l1, type(l1)))
+
+    if not isinstance(l2, list):
+        raise AnsibleFilterError('Second argument for lists_mergeby must be list. %s is %s' %
```
Comment (felixfontein) [inline plugins/filter/list.py:23]:
```suggestion
        raise AnsibleFilterError('Second argument for community.general.lists_mergeby must be list. %s is %s' %
```

File: plugins/filter/list.py [hunk 9]
```diff
@@ -0,0 +1,47 @@
...
+                                 (l2, type(l2)))
+
+    if not isinstance(index, string_types):
+        raise AnsibleFilterError('Third argument for lists_mergeby must be string. %s is %s' %
```
Comment (felixfontein) [inline plugins/filter/list.py:27]:
```suggestion
        raise AnsibleFilterError('Third argument for community.general.lists_mergeby must be string. %s is %s' %
```

File: plugins/filter/list.py [hunk 10]
```diff
@@ -0,0 +1,47 @@
...
+                                 (index, type(index)))
+
+    d = defaultdict(dict)
+    for l in (l1, l2):
+        for elem in l:
+            if index in elem.keys():
```
Comment (felixfontein) [inline plugins/filter/list.py:33]:
```suggestion
            if not instanceof(elem, Mapping):
                        raise AnsibleFilterError('Elements of list arguments for lists_mergeby must be dictionaries. Found {0!r}.'.format(elem))
            if index in elem.keys():
```

File: plugins/filter/list.py [hunk 11]
```diff
@@ -0,0 +1,47 @@
...
+                d[elem[index]].update(elem)
+    if d.values():
+        return sorted(d.values(), key=itemgetter(index))
+    else:
+        return d.values()
```
Comment (felixfontein) [inline plugins/filter/list.py:38]: To actually get a list:
```suggestion
        return []
```
But: why not simply return `sorted(d.values(), key=itemgetter(index))` in both cases? Would make the code a lot cleaner.

File: plugins/filter/list.py [hunk 12]
```diff
@@ -0,0 +1,48 @@
...
+        raise AnsibleFilterError('Second argument for community.general.lists_mergeby must be list. %s is %s' %
+                                 (l2, type(l2)))
+
+    if not isinstance(index, string_types):
+        raise AnsibleFilterError('Third argument for community.general.lists_mergeby must be string. %s is %s' %
+                                 (index, type(index)))
+
+    d = defaultdict(dict)
+    for l in (l1, l2):
+        for elem in l:
```
Comment (felixfontein) [inline plugins/filter/list.py:33]:
```suggestion
        for elem in l:
            if not instanceof(elem, Mapping):
                        raise AnsibleFilterError('Elements of list arguments for lists_mergeby must be dictionaries. Found {0!r}.'.format(elem))
```

File: plugins/filter/list.py [hunk 13]
```diff
@@ -0,0 +1,50 @@
...
+            if not instanceof(elem, Mapping):
```
Comment (felixfontein) [inline plugins/filter/list.py:34]:
```suggestion
            if not isinstance(elem, Mapping):
```
Damn, too much thinking of Java :D

File: plugins/filter/list.py [hunk 14]
```diff
@@ -0,0 +1,50 @@
...
+                        raise AnsibleFilterError('Elements of list arguments for lists_mergeby must be dictionaries. Found {0!r}.'.format(elem))
```
Comment (felixfontein) [inline plugins/filter/list.py:35]:
```suggestion
                raise AnsibleFilterError('Elements of list arguments for lists_mergeby must be dictionaries. Found {0!r}.'.format(elem))
```

**Review by felixfontein — APPROVED** [review]

File: changelogs/fragments/604-lists_mergeby-new-filter.yml [hunk 15]
```diff
@@ -0,0 +1,23 @@
+---
+minor_changes:
+  - |
+    A new filter ``lists_mergeby`` to merge two lists by an attribute.
```
Comment (dmsimard) [inline changelogs/fragments/604-lists_mergeby-new-filter.yml:4]:
```suggestion
    A new filter ``lists_mergeby`` to merge two lists of dictionaries by an attribute.
```

**Comment by dmsimard** [comment]
Just a minor, minor comment otherwise LGTM.

There's opportunities for increased test coverage by pulling in tests that @imjoseangel wrote in https://github.com/ansible-collections/community.general/pull/263/files#diff-7e7d51ef11ac4fbb4ab13c2d828552afb750ddfe6f5f04057394fd1d1833e8f7 but we can do that in a new PR.

**Review by dmsimard — APPROVED** [review]

**Comment by dmsimard** [comment]
LGTM, thanks for working on this @vbotka !

**Comment by vbotka** [comment]
> There's opportunities for increased test coverage by pulling in tests that @imjoseangel wrote in https://github.com/ansible-collections/community.general/pull/263/files#diff-7e7d51ef11ac4fbb4ab13c2d828552afb750ddfe6f5f04057394fd1d1833e8f7 but we can do that in a new PR.

I'd like to proceed here and fit the test to the filter.  What what would you prefer? Reopen this PR or let me create a new one?

**Comment by dmsimard** [comment]
> > There's opportunities for increased test coverage by pulling in tests that @imjoseangel wrote in https://github.com/ansible-collections/community.general/pull/263/files#diff-7e7d51ef11ac4fbb4ab13c2d828552afb750ddfe6f5f04057394fd1d1833e8f7 but we can do that in a new PR.
> 
> I'd like to proceed here and fit the test to the filter. What what would you prefer? Reopen this PR or let me create a new one?

@vbotka it wasn't a blocker to merge the PR so I went ahead and did that -- please open a new PR for adding new tests.

**Pull request merged**